src/services/completions.ts

```diff
@@ -739,8 +739,25 @@ namespace ts.Completions {
             }
         }
 
+        // Before offering up a JSX attribute snippet, ensure that we aren't potentially completing
+        // a tag name; this may appear as an attribute after the "<" when the tag has not yet been
+        // closed, as in:
+        //
+        //     return <>
+        //         foo <butto|
+        //     </>
+        //
+        // We can detect this case by checking if both:
+        //
+        //     1. The location is "<", so we are completing immediately after it.
+        //     2. The "<" has the same position as its parent, so is not a binary expression.
         const kind = SymbolDisplay.getSymbolKind(typeChecker, symbol, location);
-        if (kind === ScriptElementKind.jsxAttribute && preferences.includeCompletionsWithSnippetText && preferences.jsxAttributeCompletionStyle && preferences.jsxAttributeCompletionStyle !== "none") {
+        if (
+            kind === ScriptElementKind.jsxAttribute

```

Comment:
Why/how is `kind` set to `jsxAttribute` when the completion is actually a tag name? Isn’t that an issue itself, which, if fixed, would make the condition as written previously correct?

Reply:
When I last spoke to Jake, it seemed like this is an issue that other parts of the surrounding code handle similarly. A deeper fix might be something for a follow-up issue.

Reply:
I'm super AFK, but I completely agree that the right fix is to make the parser recover this differently. I was just trying to get something quick for the VS Code recovery release before I dipped out for my vacation. There's at least one other explainer comment elsewhere (some other block that is also checking that the kind is JSX attribute; I doubt I could find it from my phone!) that documents that the first attribute might be the tag name, which is why I was maybe a little less weirded out than someone looking at this PR in a vacuum.